### PR TITLE
[LOGGING] Optionally print backtrace on segfault

### PR DESCRIPTION
**Pull request by tkonolige**
Add BACKTRACE_ON_SEGFAULT cmake option to install a signal handler to print a backtrace on segfault. Disabled by default because existing signal handlers may be overriden by this signal handler. Creating a backtrace may allocate, which should not happen in a signal handler, but the program is crashing so it shouldn't make any difference.

Hopefully this will help diagnose segfaults like in #12955.

@areusch @driazati (not sure who else to tag)

### CODE REVIEW

File: src/runtime/logging.cc [hunk 1]
```diff
@@ -117,6 +121,21 @@ int BacktraceFullCallback(void* data, uintptr_t pc, const char* filename, int li
...
+void backtrace_handler(int sig) {
+  // Technically we shouldn't do any allocation in a signal handler, but
+  // Backtrace may allocate. What's the worst it could do? We're already
+  // crashing.
+  std::cerr << "!!!!!!! TVM encountered a Segfault !!!!!!!\n" << Backtrace() << std::endl;
+  exit(1);
+}
+
+__attribute__((constructor)) void install_signal_handler(void) {
+  // this may override already install signal handlers
```
Comment (driazati) [inline src/runtime/logging.cc:135]:
```suggestion
  // this may override already installed signal handlers
```

File: src/runtime/logging.cc [hunk 2]
```diff
@@ -117,6 +121,21 @@ int BacktraceFullCallback(void* data, uintptr_t pc, const char* filename, int li
...
   return 0;
 }
+
+#if TVM_BACKTRACE_ON_SEGFAULT
+void backtrace_handler(int sig) {
+  // Technically we shouldn't do any allocation in a signal handler, but
+  // Backtrace may allocate. What's the worst it could do? We're already
+  // crashing.
+  std::cerr << "!!!!!!! TVM encountered a Segfault !!!!!!!\n" << Backtrace() << std::endl;
+  exit(1);
```
Comment (driazati) [inline src/runtime/logging.cc:131]: we can at least call back to the default signal handler with this to get the standard `terminated by signal SIGSEGV` text

```suggestion

  struct sigaction act;
  memset(&act, 0, sizeof(struct sigaction));
  act.sa_flags = SA_RESETHAND;
  act.sa_handler = SIG_DFL;
  sigaction(sig, &act, NULL);
  raise(sig);
```

File: tests/scripts/task_config_build_cpu.sh [hunk 3]
```diff
@@ -55,6 +55,7 @@ echo set\(USE_CMSISNN OFF\) >> config.cmake
 echo set\(USE_VITIS_AI ON\) >> config.cmake
 echo set\(USE_VERILATOR ON\) >> config.cmake
 echo set\(USE_LIBBACKTRACE ON\) >> config.cmake
+echo set\(BACKTRACE_ON_SEGFAULT ON\) >> config.cmake
```
Comment (driazati) [inline tests/scripts/task_config_build_cpu.sh:58]: What do you think about adding it to every CI config (i.e. all the `task_*` files)? We often get crashes in the cortexm and hexagon builds too

**Review by driazati — APPROVED** [review]
this is great! This is something we've been lacking for a while, I had tried a few times to get core dumps from segfaulting CI jobs but never could get it working in Docker, but this fills the same gap

File: tests/scripts/task_config_build_cpu.sh (hunk 3 above)
Comment (tkonolige) [reply tests/scripts/task_config_build_cpu.sh:58]: Will do!

**Pull request merged**